src/languages/swift.js

```diff
@@ -283,6 +288,143 @@ export default function(hljs) {
   };
   TYPE.contains.push(GENERIC_ARGUMENTS);
 
+  // https://docs.swift.org/swift-book/ReferenceManual/Expressions.html#ID552
+  // Prevents element names from being highlighted as keywords.
+  const TUPLE_ELEMENT_NAME = {
+    begin: lookahead(concat(Swift.identifier, /\s*:/)),
+    end: lookahead(/:/),
+    keywords: "_",
+    relevance: 0
+  };
+  // Matches tuples as well as the parameter list of a function type.
+  const TUPLE = {
+    begin: /\(/,
+    end: /\)/,
+    relevance: 0,
+    keywords: KEYWORDS,
+    contains: [
+      'self',
+      TUPLE_ELEMENT_NAME,
+      ...COMMENTS,
+      ...KEYWORD_MODES,
+      ...BUILT_INS,
+      ...OPERATORS,
+      NUMBER,
+      STRING,
+      ...IDENTIFIERS,
+      ...ATTRIBUTES,
+      TYPE
+    ]
+  };
+
+  // https://docs.swift.org/swift-book/ReferenceManual/Declarations.html#ID362
+  // Matches both the keyword func and the function title.
+  // Grouping these lets us differentiate between the operator function <
+  // and the start of the generic parameter clause (also <).
+  const FUNCTION_TITLE = {
+    begin: lookahead(/\bfunc\b/),
+    relevance: 0,
+    contains: [
+      {
+        className: 'keyword',
+        begin: /\bfunc\b/
+      },
+      {
+        begin: /\s+/,
+        relevance: 0
+      },
+      {
+        className: 'title',
+        begin: either(QUOTED_IDENTIFIER.begin, Swift.identifier, Swift.operator),
+        // Required to make sure the opening < of the generic parameter clause
+        // isn't parsed as a second title.
+        endsParent: true,
+        relevance: 0
+      }
+    ]
+  };
+  const GENERIC_PARAMETERS = {
+    begin: /\s*</,
```

Comment:
This \s* looks out of place and should probably be handled elsewhere.

Reply:
This is because there's no `end` rule anymore, so the spaces have to be matched in order for the `FUNCTION` mode to continue. That's why I added optional leading spaces in a few places.

Reply:
Yes, I understand why.  They should be inside the FUNCTION mode though not appended to the child modes.

Reply:
I think those leading spaces are just a leftover:

- The original version of this grammar combined a recursive `params` rule with `endsParent`, which caused https://github.com/highlightjs/highlight.js/issues/2857.
- My first fix was to keep the recursion, but remove `endsParent`.
- However, I then no longer had an end condition, so I had to make sure the rules kept matching up until the end of the parameter list.
- This meant I had to allowed leading spaces in certain places, but not just whitespace anywhere. Otherwise, a function could have two parameter lists for example.
- However, thanks to the `TUPLE` rule, I no longer need `contains: 'self'` for function parameters, so I can use `endsParent` again and remove the spaces.

So be warned, removing the `TUPLE` rule will break all of this again, as you will have to go back to recursive `FUNCTION_PARAMETERS`, which means you can't use `endsParent` anymore. You also can't use the `end: '{'` rule from the original version, as that was one of the causes for https://github.com/highlightjs/highlight.js/issues/2895

So long story short, the main motivation for `TUPLE` is to allow nested **parentheses** in `FUNCTION_PARAMETERS`, without nesting `FUNCTION_PARAMETERS` itself.

Reply:
> Otherwise, a function could have two parameter lists for example.

If this never happens in real Swift code (why would a function have two param lists) then it's not a concern we need to worry about.  Perhaps I misunderstand what you're saying though. 

> So long story short, the main motivation for TUPLE is to allow nested parentheses in FUNCTION_PARAMETERS, without nesting FUNCTION_PARAMETERS itself.

Ah... or else `)` will potentially end everything prematurely... often a placeholder recursive rule is used for this sort of them, but then if it needed all the same children might as well just use `FUNCTION_PARAMETERS` as you're thinking.  And TUPLE might make more sense than that since it provides a bit of addl value.

Reply:
Yeah, nesting `FUNCTION_PARAMETERS` within itself will also highlights identifiers in tuples and function types as `params`, which is what I wanted to avoid, as those aren't highlighted elsewhere, and aren't `params` anyway.

I also wanted to narrow down the scope of `params` to just the identifiers, and not the entire parameter list, as that causes overwhelmingly large highlights, and also highlights all the punctuation in the parameter list (and everything else we don't normally highlight).